### ISSUE TITLE
2-platform clone doesn't work

### ISSUE THREAD

**Issue by khaeru** [iiasa/message_ix#198]
*mea culpa*: #182 did not entirely fix #174, it seems. I run the following code with `gp3.properties` pointing to IIASA GP3:

```python
from pathlib import Path

import ixmp
import message_ix

db_path = Path('~', '.local', 'share', 'ixmp', 'localdb', 'bug').expanduser()

def info(s):
    print(s.par('input').head(), s.set('node'), sep='\n')


# 1. Load remote scenario

mp1 = ixmp.Platform('gp3.properties')
s1 = message_ix.Scenario(mp1, 'CD_Links_SSP2_v2', 'baseline')

info(s1)


# 2. Clone to a local database

mp2 = ixmp.Platform(db_path, dbtype='HSQLDB')  # (A)
s2 = s1.clone(platform=mp2)                    # (B)

print(mp2.scenario_list(default=False))

info(s2)

# Clean up
mp1.close_db()
mp2.close_db()
del mp1, mp2, s1, s2


# 3. Load from the local database

mp3 = ixmp.Platform(db_path, dbtype='HSQLDB')

print(mp3.scenario_list(default=False))

s3 = message_ix.Scenario(mp3, 'CD_Links_SSP2_v2', 'baseline')

info(s3)
```

This gives the following output. Note that the `input` parameter is **empty** in the cloned scenario; this is the bug.

<details><summary>Console output…</summary>

```
$ python3 bug.py
INFO:root:launching ixmp.Platform using config file at '/home/khaeru/.local/share/ixmp/gp3.properties'
2019-05-28 14:44:25,360  INFO at.ac.iiasa.ixmp.Platform:117 - Welcome to the IX modeling platform!
2019-05-28 14:44:25,368  INFO at.ac.iiasa.ixmp.Platform:118 -  connected to database 'jdbc:oracle:thin:@gp3.iiasa.ac.at:1521:GP3' (user: ene_ixmp)...                                                                                                         
2019-05-28 14:44:25,432  INFO at.ac.iiasa.ixmp.objects.Scenario:208 - loading default Scenario 'CD_Links_SSP2_v2|baseline' (version: 4, runid: 21779)...                                                                                                      
2019-05-28 14:44:25,816  INFO at.ac.iiasa.ixmp.objects.Scenario:234 - done loading Scenario from the database!
2019-05-28 14:44:25,885  INFO at.ac.iiasa.ixmp.objects.MsgScenario:451 -  spatial and temporal disaggregation maps assignment                                                                                                                                 
2019-05-28 14:44:25,902  INFO at.ac.iiasa.ixmp.objects.MsgScenario:567 -  period mapping assignment                                                                                                                                                           
2019-05-28 14:44:25,910  INFO at.ac.iiasa.ixmp.objects.MsgScenario:593 -  duration period assignment
2019-05-28 14:44:25,924  INFO at.ac.iiasa.ixmp.objects.Scenario:1945 - committing changes of Scenario 'CD_Links_SSP2_v2|baseline' to the database (runid: 21779)...                                                                                           
2019-05-28 14:44:27,885  INFO at.ac.iiasa.ixmp.objects.TimeSeries:813 - done updating Scenario 'CD_Links_SSP2_v2|baseline' to the database (runid: 21779, version: 4)!                                                                                        
2019-05-28 14:44:27,895  INFO at.ac.iiasa.ixmp.objects.MsgScenario:451 -  spatial and temporal disaggregation maps assignment                                                                                                                                 
2019-05-28 14:44:27,896  INFO at.ac.iiasa.ixmp.objects.MsgScenario:567 -  period mapping assignment                                                                                                                                                           
2019-05-28 14:44:27,897  INFO at.ac.iiasa.ixmp.objects.MsgScenario:593 -  duration period assignment                                                                                                                                                          
  node_loc technology  year_vtg  year_act mode node_origin commodity      level  time time_origin  value unit
0  R11_PAO    eth_bio      1995      1995   M1     R11_PAO    electr  secondary  year        year    0.0  GWa
1  R11_PAO    eth_bio      2000      2000   M1     R11_PAO    electr  secondary  year        year    0.0  GWa
2  R11_PAO    eth_bio      2005      2005   M1     R11_PAO    electr  secondary  year        year    0.0  GWa
3  R11_PAO    eth_bio      2010      2010   M1     R11_PAO    electr  secondary  year        year    0.0  GWa
4  R11_PAO    eth_bio      2020      2020   M1     R11_PAO    electr  secondary  year        year    0.0  GWa
0       World
1     R11_AFR
2     R11_CPA
3     R11_EEU
4     R11_FSU
5     R11_GLB
6     R11_LAM
7     R11_MEA
8     R11_NAM
9     R11_PAO
10    R11_PAS
11    R11_SAS
12    R11_WEU
dtype: object
INFO:root:launching ixmp.Platform with local HSQLDB database at '/home/khaeru/.local/share/ixmp/localdb/bug'
2019-05-28 14:44:33,431  INFO at.ac.iiasa.ixmp.Platform:117 - Welcome to the IX modeling platform!
2019-05-28 14:44:33,432  INFO at.ac.iiasa.ixmp.Platform:118 -  connected to database 'jdbc:hsqldb:file:/home/khaeru/.local/share/ixmp/localdb/bug' (user: ixmp)...                                                                                            
2019-05-28 14:45:42,634  INFO at.ac.iiasa.ixmp.objects.Scenario:2015 - saving Scenario 'CD_Links_SSP2_v2|baseline' to database (runid: 0)...                                                                                                                  
2019-05-28 14:45:44,772  INFO at.ac.iiasa.ixmp.objects.TimeSeries:816 - done saving Scenario 'CD_Links_SSP2_v2|baseline' to the database (runid: 0, version: 1)!                                                                                              
2019-05-28 14:45:45,400  INFO at.ac.iiasa.ixmp.objects.Scenario:211 - loading Scenario 'CD_Links_SSP2_v2|baseline' by version id (version: 1, runid: 0)...                                                                                                    
2019-05-28 14:45:45,434  INFO at.ac.iiasa.ixmp.objects.Scenario:234 - done loading Scenario from the database!
2019-05-28 14:45:45,435  INFO at.ac.iiasa.ixmp.objects.MsgScenario:451 -  spatial and temporal disaggregation maps assignment
2019-05-28 14:45:45,436  INFO at.ac.iiasa.ixmp.objects.MsgScenario:567 -  period mapping assignment
2019-05-28 14:45:45,437  INFO at.ac.iiasa.ixmp.objects.MsgScenario:593 -  duration period assignment
              model  scenario   scheme is_default is_locked cre_user                    cre_date upd_user upd_date lock_user lock_date                                         annotation  version                                                            
0  CD_Links_SSP2_v2  baseline  MESSAGE      false     false   khaeru  2019-05-28 14:45:42.634000     None     None      None      None  clone Scenario from 'CD_Links_SSP2_v2|baseline...        1                                                            
Empty DataFrame
Columns: [node_loc, technology, year_vtg, year_act, mode, node_origin, commodity, level, time, time_origin, value, unit]
Index: []
0       World
1     R11_AFR
2     R11_CPA
3     R11_EEU
4     R11_FSU
5     R11_GLB
6     R11_LAM
7     R11_MEA
8     R11_NAM
9     R11_PAO
10    R11_PAS
11    R11_SAS
12    R11_WEU
dtype: object
2019-05-28 14:45:45,470  INFO at.ac.iiasa.ixmp.Platform:148 - closed the connection to database 'jdbc:oracle:thin:@gp3.iiasa.ac.at:1521:GP3'                                                                                                                  
2019-05-28 14:45:46,020  INFO at.ac.iiasa.ixmp.Platform:148 - closed the connection to database 'jdbc:hsqldb:file:/home/khaeru/.local/share/ixmp/localdb/bug'                                                                                                 
INFO:root:launching ixmp.Platform with local HSQLDB database at '/home/khaeru/.local/share/ixmp/localdb/bug'
2019-05-28 14:45:47,262  INFO at.ac.iiasa.ixmp.Platform:117 - Welcome to the IX modeling platform!
2019-05-28 14:45:47,262  INFO at.ac.iiasa.ixmp.Platform:118 -  connected to database 'jdbc:hsqldb:file:/home/khaeru/.local/share/ixmp/localdb/bug' (user: ixmp)...                                                                                            
              model  scenario   scheme is_default is_locked cre_user                    cre_date upd_user upd_date lock_user lock_date                                         annotation  version                                                            
0  CD_Links_SSP2_v2  baseline  MESSAGE      false     false   khaeru  2019-05-28 14:45:42.634000     None     None      None      None  clone Scenario from 'CD_Links_SSP2_v2|baseline...        1                                                            
2019-05-28 14:45:47,280  INFO at.ac.iiasa.ixmp.objects.Scenario:211 - loading Scenario 'CD_Links_SSP2_v2|baseline' by version id (version: 1, runid: 0)...                                                                                                    
2019-05-28 14:45:47,305  INFO at.ac.iiasa.ixmp.objects.Scenario:234 - done loading Scenario from the database!
2019-05-28 14:45:47,307  INFO at.ac.iiasa.ixmp.objects.MsgScenario:451 -  spatial and temporal disaggregation maps assignment
2019-05-28 14:45:47,307  INFO at.ac.iiasa.ixmp.objects.MsgScenario:567 -  period mapping assignment
2019-05-28 14:45:47,308  INFO at.ac.iiasa.ixmp.objects.MsgScenario:593 -  duration period assignment
Empty DataFrame
Columns: [node_loc, technology, year_vtg, year_act, mode, node_origin, commodity, level, time, time_origin, value, unit]
Index: []
0       World
1     R11_AFR
2     R11_CPA
3     R11_EEU
4     R11_FSU
5     R11_GLB
6     R11_LAM
7     R11_MEA
8     R11_NAM
9     R11_PAO
10    R11_PAS
11    R11_SAS
12    R11_WEU
dtype: object
```
</details>

Other information:
- Note that the test added in #182 is between two local databases. I have no idea why this would make a difference.
- `bug.lobs` in the database directory is < 50 kB after the line marked `(A)` and > 500 kB after the line marked `(B)`. And the `bug.script` shows many INSERT lines. @zikolach is it possible the data is getting inserted but then can't be located again, for some reason? I'm familiar with SQL but not the database schema.

cc @gidden @danielhuppmann

**Comment by zikolach** [comment]
@khaeru can you upload hsqldb database (archive) somewhere to look at it?

**Comment by gidden** [comment]
Out of curiosity @khaeru, what is the result for each Scenario if you print `scen1.set('node')` and `scen1.set('year')` since those are the two calls being unit tested?

**Comment by khaeru** [comment]
> @khaeru can you upload hsqldb database (archive) somewhere to look at it?

Done, see my message in Slack.

> Out of curiosity @khaeru, what is the result for each Scenario if you print `scen1.set('node')` and `scen1.set('year')` since those are the two calls being unit tested?

Good question. Updated the description and console output with the result: sets are available from the cloned scenario, even though parameter data is not.

**Comment by gidden** [comment]
Welp, that sounds crazy. I'll let @zikolach and @danielhuppmann chime in from here..

**Comment by gidden** [comment]
Maybe its worth adding a quick check in the existing unit test? You could add a parameter and then check it in the clone to see if its indeed a generic problem or specifically oracle-local problem.

**Comment by khaeru** [comment]
> Maybe its worth adding a quick check in the existing unit test? You could add a parameter and then check it in the clone to see if its indeed a generic problem or specifically oracle-local problem.

Appears to be a generic problem. Here is a modified version of `test_core.test_clone` that can be used; it fails at “Data can be retrieved from clone”.

```py
def test_clone(tmpdir):
    # Two local platforms
    mp1 = ixmp.Platform(tmpdir / 'mp1', dbtype='HSQLDB')
    mp2 = ixmp.Platform(tmpdir / 'mp2', dbtype='HSQLDB')

    # A minimal scenario
    scen1 = Scenario(mp1, model='model', scenario='scenario', version='new')
    scen1.add_spatial_sets({'country': 'Austria'})
    scen1.add_set('technology', 'bar')
    scen1.add_horizon({'year': [2010, 2020]})

    # ADDED: Store data
    mp1.add_unit('$')
    data = pd.DataFrame(dict(node_loc='Austria', technology='bar',
                             year_vtg=2010, value=1.2, unit='$'), index=[0])
    scen1.add_par('inv_cost', data)
    pdt.assert_frame_equal(scen1.par('inv_cost').astype({'value': float}),
                           data)

    scen1.commit('add minimal sets for testing')

    assert len(mp1.scenario_list(default=False)) == 1

    # Clone
    scen2 = scen1.clone(platform=mp2)

    # Return type of ixmp.Scenario.clone is message_ix.Scenario
    assert isinstance(scen2, Scenario)

    # ADDED: Data can be retrieved from clone
    pdt.assert_frame_equal(scen2.par('inv_cost'), data)

    # Close and re-open both databases
    mp1.close_db()  # TODO this should be done automatically on del
    mp2.close_db()  # TODO this should be done automatically on del
    del mp1, mp2
    mp1 = ixmp.Platform(tmpdir / 'mp1', dbtype='HSQLDB')
    mp2 = ixmp.Platform(tmpdir / 'mp2', dbtype='HSQLDB')

    # Same scenarios present in each database
    assert all(mp1.scenario_list(default=False) ==
               mp2.scenario_list(default=False))

    # Load both scenarios
    scen1 = Scenario(mp1, 'model', 'scenario')
    scen2 = Scenario(mp2, 'model', 'scenario')

    # Contents are identical
    assert all(scen1.set('node') == scen2.set('node'))
    assert all(scen1.set('year') == scen2.set('year'))

    # ADDED
    pdt.assert_frame_equal(scen1.par('inv_cost'), scen2.par('inv_cost'))
```

**Comment by danielhuppmann** [comment]
this should be fixed with [ixmp release 0.2.0](https://github.com/iiasa/ixmp/releases/tag/v0.2.0), unit tests have been added to `message_ix` to prevent regression
